moshi/moshi/models/tts.py

```diff
@@ -0,0 +1,624 @@
+# Copyright (c) Kyutai, all rights reserved.
+"""
+Implements the logic for the state machine around the Delayed Streams Modeling (DSM) based TTS model.
+"""
+
+from collections import deque
+from dataclasses import dataclass, field
+from functools import cached_property
+import re
+from pathlib import Path
+import typing as tp
+
+from safetensors.torch import load_file
+from sentencepiece import SentencePieceProcessor
+import sphn
+import torch
+
+from ..conditioners import ConditionAttributes, dropout_all_conditions, TensorCondition
+from ..conditioners.text import LUTConditioner
+from . import loaders, MimiModel, LMModel, LMGen
+
+
+DEFAULT_DSM_TTS_REPO = "alex-testing/dsm-tts-rc"
+DEFAULT_DSM_TTS_VOICE_REPO = 'alex-testing/dsm-tts-rc-voices'
+
+
+@dataclass
+class TokenIds:
+    """
+    The token ids for special tokens:
+        - word: a new word is starting.
+        - pad: padding, nothing happens.
+        - main: indicates the start of turn of the main speaker.
+        - other: indicates the start of turn of the other speaker.
+        - zero: special value that is embedded to exactly 0.
+        - ungenerated: indicate that a value is not yet generated but should be
+        - card: text cardinality, including the initial token (1 + tokenizer cardinality).
+            This is used for multiplexing multiple input tokens into the text stream.
+
+    """
+    word = 0
+    pad = 3
+    main = 1
+    other = 2
+    zero = -1
+    ungenerated = -2
+    card: int = 8001
+
+
+@dataclass
+class Entry:
+    """One word to generate.
+
+    Args:
+        tokens: list of tokens for this word.
+        word: word as string.
+        padding: if > 0, we will force that many padding after the word.
+            To have an effect this should be more than len(tokens), as we will
+            prevent new word until we have passed all the text tokens.
```

Comment:
The wording here is a bit confusing to me, on one hand it says *padding after the word* but on the other end `len(tokens)` seems to be included as part of padding. Maybe this could be clarified?